### PR TITLE
feat(FN-1685): fix csv parsing for rows with empty data at end of row

### PR DESCRIPTION
**Pull request by ChristianMcc**
## Introduction :pencil2:
Empty cells at the end of a row would not be processed by the excel library and so we would not get the cell addresses or empty values for these cells when processing

## Resolution :heavy_check_mark:
Keep count of how many header columns we have and add in any empty values with cell addresses so we can show descriptive error messages later on in the process.

### CODE REVIEW

File: portal/server/utils/csv-utils.js [hunk 1]
```diff
@@ -13,6 +13,15 @@ const columnIndexToExcelColumn = (index) => {
   return result;
 };
 
+const excelColumnToColumnIndex = (column) => {
```
Comment (toddym42) [inline portal/server/utils/csv-utils.js:16]: Could we add a JSDoc with the type of `column` and the return type.

File: portal/server/utils/csv-utils.js [hunk 2]
```diff
@@ -13,6 +13,15 @@ const columnIndexToExcelColumn = (index) => {
...
+  let result = 0;
+  for (let i = 0; i < column.length; i += 1) {
+    result *= 26;
+    result += column.charCodeAt(i) - 64;
```
Comment (toddym42) [inline portal/server/utils/csv-utils.js:20]: Could we give some names to the `26` and `64` constants to explain what they are?

File: portal/server/utils/csv-utils.js [hunk 3]
```diff
@@ -43,11 +56,31 @@ const parseXlsxToCsvArrays = (worksheet) => {
...
         rowDataWithCellAddresses.push(`${cellValue}`);
+        headerCount += 1;
       } else {
         rowDataWithCellAddresses.push(`${cellValue}-${cell.address}`);
+        lastAddress = cell.address;
+        columnCount += 1;
       }
     });
+
+    if (firstRow === false) {
```
Comment (toddym42) [inline portal/server/utils/csv-utils.js:67]: Can just be `if (!firstRow)`

File: portal/server/utils/csv-utils.js [hunk 4]
```diff
@@ -43,11 +56,31 @@ const parseXlsxToCsvArrays = (worksheet) => {
...
+      // If the row has no data in the final columns of the row, we need to fill in the empty cells with
+      // empty strings and cell addresses so that the csv parser library can parse the data correctly
+      while (columnCount < headerCount) {
+        rowData.push('');
+        // calculate the cell address here given the last address
+        const lastAddressMatch = lastAddress.match(CELL_ADDRESS_REGEX);
+        const [lastColumn, lastRow] = lastAddressMatch.slice(1);
+        const lastColumnIndex = excelColumnToColumnIndex(lastColumn);
+        const newColumn = columnIndexToExcelColumn(parseInt(lastColumnIndex, 10) + 1);
```
Comment (toddym42) [inline portal/server/utils/csv-utils.js:76]: If we give a (JSDoc) return type of `excelColumnToColumnIndex` I think we can remove `parseInt` here?

File: portal/server/utils/csv-utils.test.js [hunk 5]
```diff
@@ -21,6 +22,20 @@ describe('csv-utils', () => {
...
+      const columnIndex = excelColumnToColumnIndex('C');
+
+      expect(columnIndex).toBe(2);
+    });
+
+    it('returns the correct index for a column above 26', async () => {
+      const columnIndex = excelColumnToColumnIndex('AE');
+
+      expect(columnIndex).toBe(30);
+    });
```
Comment (toddym42) [inline portal/server/utils/csv-utils.test.js:36]: What's the significance of 26 here?

File: portal/server/utils/csv-utils.test.js [hunk 6]
```diff
@@ -21,6 +22,20 @@ describe('csv-utils', () => {
...
+  });
+
```
Comment (toddym42) [inline portal/server/utils/csv-utils.test.js:38]: Could we add any tests to check the new logic does what we expect (adding empty strings to empty cells)

File: portal/server/utils/csv-utils.test.js (hunk 5 above)
Comment (ChristianMcc) [reply portal/server/utils/csv-utils.test.js:36]: This just represents wrapping from a single character column code to multiple characters. I can expand the test case name to include this if that's helpful?

File: portal/server/utils/csv-utils.test.js (hunk 6 above)
Comment (ChristianMcc) [reply portal/server/utils/csv-utils.test.js:38]: Mm yes I definitely should have just added these, will add something :)

File: portal/server/utils/csv-utils.test.js (hunk 5 above)
Comment (toddym42) [reply portal/server/utils/csv-utils.test.js:36]: Ah, of course. Probably wouldn't hurt to be a bit more explicit

File: portal/server/utils/csv-utils.js (hunk 2 above)
Comment (ChristianMcc) [reply portal/server/utils/csv-utils.js]: Added some comments, haven't named as thought comments would do a better job not sure if that's a red flag.
26 = number of characters in alphabet
64 = A is the 65th Unicode character so making A = 1

File: portal/server/utils/csv-utils.js [hunk 7]
```diff
@@ -3,6 +3,18 @@ const csv = require('csv-parser');
 const { Readable } = require('stream');
 const { CELL_ADDRESS_REGEX } = require('../constants/regex');
 
+/**
+ * @typedef {import('exceljs').Worksheet} Worksheet
+ * @typedef {Object} parsedXlsxDataResponse
```
Comment (toddym42) [inline portal/server/utils/csv-utils.js:8]: Nit: types would usually start with an upper-case. i.e. `ParsedXlsxDataResponse`

File: portal/server/utils/csv-utils.js [hunk 8]
```diff
@@ -3,6 +3,18 @@ const csv = require('csv-parser');
...
+ * @property {Object} csvData - The index of the object
+ * @property {Object} csvDataWithCellAddresses - The error message of the object
```
Comment (toddym42) [inline portal/server/utils/csv-utils.js:10]: Looks like a copy-pasta on the descriptions here.

**Review by toddym42 — APPROVED** [review]
